text/0000-poll-drop-ready.md

```diff
@@ -0,0 +1,480 @@
+field of the type. The order of operations would be:
+
+1. `poll_drop_ready` runs, pin projecting to each field of the type.
+2. The type's destructor runs, potentially moving each field.
+3. Each field's destructor runs, after the type's destructor has potentially
+moved them.
+
+Because of this sequence, the in projection in `poll_drop_ready` would be
+unsound. Therefore, `poll_drop_ready` cannot take self by Pin. Here is a bit of
+code demonstrating what I mean:
+
+```rust
+// If `field` implements `poll_drop_ready` and `poll_drop_ready` used `Pin`,
+// `field` would be witnessed as pinned before calling drop on `Foo`.
+struct Foo<T: Default> {
+    field: T
+}
+
+impl<T: Default> Drop for Foo<T> {
+    fn drop(&mut self) {
+        // Then, in the destructor for `Foo`, we can move `field` and re-pin it
+        // at a different location, violating the guarantees of `Pin`.
+
+        let moved_field = mem::replace(&mut self.field, T::default());
+        let pinned: Box<Pin<T>> = Box::pin(moved_field);
+    }
+}
+```
+
+However, its worth noting that a user can locally reason about whether private
+fields can be moved or accessed again, based on the APIs exposed by a type,
+and, if necessary, reconstruct a pin of those fields. [The pin documentation
+contains relevant information.][pin-docs] In other words, if your drop
+implementation is implemented correctly, you can construct a `Pin` of self
+inside your `poll_drop_ready`.
+
+## Drop order
+
+The order of calls to `poll_drop_ready` can be inferred from the rest of this
+document, but it's good to specify them explicitly.
+
+Between variables, calls to `poll_drop_ready` will occur in reverse order of
+the variable's introduction, immediately prior to the calls to that variable's
+destructor.
+
+Between fields of a type, calls to `poll_drop_ready` will occur in the textual
+order of that type's declaration, with the call for the type itself occuring
+first. This is similar to the order for calls to drop. *However*, these calls
+will occur in a loop, at the level of the type being dropped, until all calls
+to `poll_drop_ready` for that value have returned `Ready`, so they will be
+"interleaved" and concurrent in practice. The program will *not* wait for each
+field to return ready before beginning to process the subsequent field.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Like all language extensions which increase the expressiveness of the language,
+this feature will increase the complexity of Rust. In particular, this feature
+sits squarely at the intersection of async code and destructors, both areas of
+the language which involved nuanced guarantees about memory and correctness.
+Therefore, the biggest drawback of adding this feature is that it increases the
+complexity of the language.
+
+It should also be noted that it introduces silent "yield points" into async
+contexts, whereas previously all yield points were visible awaits. This is just
```

Comment:
Does that mean, that in an asynchronous context, a drop of non-copy temporaries at the end of statement / scope always introduces a yield point? Can it be sometimes avoided, if so on what basis? What about code that drops generic `T` temporary?

The presence or absence of yield points shapes the overall size of the generator, the determination which lifetimes and types will be included in it, how long the borrows extend, whether resulting future type will by Sync or Send, etc. Introduction of additional yield points to an existing async code base is unlikely to be backward compatible change.

Reply:
Optimizing no-op yield points away is specifically mentioned in the RFC as desirable, and should usually be achievable, maybe benefiting from MIR optimization passes. The biggest concern will be trait objects, because the call will be virtual. We may need specific representational optimizations to prevent that from being a problem.

However, you raise a very good point: this optimization would be necessary to influence the auto trait implementation of async functions, since we must *only* store state for types which have a meaningful poll_drop_ready in order to avoid a backward compatibility hazard. We probably need to figure out how we could solve that hazard.

Reply:
Actually, it may be that trait objects scupper this plan:

```rust
async fn foo(obj: Box<dyn Any>) {
    let rc = Rc::new(0);
    let _ = obj;
    drop(rc);
}
```

Can't see a way that this doesn't become a future that no longer implements `Send`. At very least, implementing this may require crater, lints, and a serious migration effort.